core/src/main/java/com/radixdlt/genesis/olympia/OlympiaEndStateApiClient.java

```diff
@@ -0,0 +1,165 @@
+ * cryptographic keys, tokens or assets created, exchanged, stored or arising from any
+ * interaction with the Work;
+ *
+ * B. any failure in a transmission or loss of any token or assets keys or other digital
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+package com.radixdlt.genesis.olympia;
+
+import com.google.inject.Inject;
+import com.radixdlt.networks.Network;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+public final class OlympiaEndStateApiClient {
+  private static final String END_STATE_API_PATH = "/olympia-end-state";
+
+  private static final String STATUS_FIELD = "status";
+  private static final String STATUS_READY = "ready";
+  private static final String STATUS_NOT_READY = "not_ready";
+
+  public sealed interface OlympiaEndStateResponse {
+    record Ready(String signature, String hash, String base64Contents)
+        implements OlympiaEndStateResponse {}
+
+    record NotReady(String signature, String placeholderHash) implements OlympiaEndStateResponse {}
+  }
+
+  private final URL endStateApiUrl;
+  private final Network network;
+
+  @Inject
+  public OlympiaEndStateApiClient(OlympiaGenesisConfig config, Network network) {
+    try {
+      this.endStateApiUrl = new URL(config.nodeCoreApiUrl(), END_STATE_API_PATH);
+    } catch (MalformedURLException e) {
+      throw new RuntimeException(
+          // Shouldn't really happen, nodeCoreApiUrl is verified earlier
+          String.format(
+              "Failed to create OlympiaEndStateApiClient. Invalid endStateApiUrl (coreApiUrl=%s"
+                  + " path=%s)",
+              config.nodeCoreApiUrl(), END_STATE_API_PATH));
+    }
+    this.network = network;
+  }
+
+  public OlympiaEndStateResponse getOlympiaEndState() throws IOException, JSONException {
```

Comment:
_(minor - I understand this is almost like a one-off script code)_

Instead of handcrafting this, would you consider one of light-weight rest clients, like https://github.com/jcabi/jcabi-http which could achieve this in ~7 lines? (remember, boilerplate code is where bugs live :P )

Reply:
I have mixed feelings about this :D
On one hand, this is a critical path so we probably shouldn't rely on 3rd party libs if it's not required
but on the other hand, this is a critical path so maybe we should use a battle-tested code instead of a handcrafted one :man_shrugging:

Reply:
I've used a library :)